### DISCUSSION TITLE
make_iterator

### DISCUSSION THREAD

**Discussion by ajum** [wjakob/nanobind#51] [Q&A] [answered]
Is it possible to add or reproduce this pybind11 functionality in nanobind?

**Comment by efifogel** [comment]
I've tried to port the implementation in pybind11 to nanobind, and immediately encountered the following problem.
The following statement appears in the pybind11 implementation:
`class_<state>(handle(), "iterator", pybind11::module_local())`
Is there an equivalent to this statement in nanobind?
I assume that the answer is negative.
So, without getting into too many details (related to lifetime of class and module wrappers), my assumption is that the best way to provide something close to `make_iterator` is to split it into two pieces:
1. Introduce a function that defines a wrapper for the iterator.
2. Add attributes to modules or classes for this iterator.

Can someone confirm my assumption?

BW, I think that the pybind11 implementation has a degeneracy; The user cannot really change `ValueType` (e.g., to be a reference type), because the `result_type` in `iterator_access` is hard coded.
Also, the name for the iterator attribute is hard coded. Splitting the functionality as described above enables the provision of a name and in turn obtaining a Python iterator explicitly using this name.

**Comment by efifogel** [comment]
Some parts are taken from pybind11.

namespace py = nanobind;

//
template <typename Type, typename PyClass>
bool add_attr(PyClass& cls, const char* name) {
  const py::handle info = py::type<Type>();
  if (! info.is_valid() || ! py::type_check(info)) return false;
  cls.attr(name) = info;
  return true;
}

//
template <typename Iterator, typename Sentinel>
struct iterator_state {
  Iterator it;
  Sentinel end;
  bool first_or_done;
};

//
template <py::rv_policy Policy,
          typename Iterator, typename Sentinel, typename ValueType,
          typename... Extra,
          typename C>
void add_iterator_impl(const char* name, C& c, Extra&&... extra) {
  using state = iterator_state<Iterator, Sentinel>;
  if (add_attr<state>(c, name)) return;

  py::class_<state>(c, name)
    .def("__iter__", [](state& s)->state& { return s; })
    .def("__next__", [](state& s) -> ValueType {
                       if (! s.first_or_done) ++s.it;
                       else s.first_or_done = false;
                       if (s.it == s.end) {
                         s.first_or_done = true;
                         throw py::stop_iteration();
                       }
                       return *s.it;
                     },
      std::forward<Extra>(extra)..., Policy)
    ;
}

// Add (wrap) an iterator
template <typename Iterator, typename Sentinel,
          typename ValueType = decltype(*std::declval<Iterator>()),
          py::rv_policy Policy = py::rv_policy::reference_internal,
          typename... Extra,
          typename C>
void add_iterator(const char* name, C& c, Extra&&... extra) {
  add_iterator_impl<Policy,
                    Iterator, Sentinel, ValueType,
                    Extra...>(name, c, std::forward<Extra>(extra)...);
}

// Obtain a Python iterator
template <typename Iterator, typename Sentinel>
py::object make_iterator(Iterator begin, Sentinel end) {
  using state = iterator_state<Iterator, Sentinel>;
  return py::cast(state{begin, end, true});
}


   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by efifogel** [comment]
What do you mean?
It's a function in nanobind defined in nb_error.h, which is included by
nanobind.h.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by jswarup** [comment]
Apologies for delay on geting back.

I missed a call to register the iterator. It works fine.
add_iterator<const double *, const double *>( name, mod);
Thanks
Jyoti

**Comment by wjakob** [comment, marked as answer]
FYI: This functionality has now been officially added to `nanobind`. However, there are small differences compared to _pybind11_. See the change to `README.md` and the test suite example for details.

https://github.com/wjakob/nanobind/commit/34d0be1bbeb54b8265456fd3a4a50e98f93fe6d4
